### PR TITLE
Check the cluster status

### PR DESCRIPTION
**Pull request by gfphoenix78**
Currently, most (if not all) test cases share the same GPDB cluster, which is much faster than creating a separate GPDB cluster for each test. But it has an obvious side-effect, some tests may break the cluster and make the cluster status undetermined for the later test cases. Some test cases are highly dependent on the assumption that all instances are up, for example, pg_basebackup_with_tablespaces. Some test cases are not sensitive to the changes. The test case itself that breaks the cluster status and some later test cases may all pass the ICW, but a later test case could fail, which makes it hard to find the root cause.
Actually, the test case that unexpectedly breaks the cluster status should also fail. There is no guarantee for it.
To achieve this goal, there should be a mechanism to do some optional checks. I added two options in pg_regress and pg_isolation_regress, `--hook-test` and `--hook-points`. `--hook-test` specifies which test case to use to check. `--hook-points` specifies after which test cases to run checks. `--hook-points` could be omitted, which means running checks after every test group.

- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
For example, there is a test case to check the cluster status in sql:
sql/check_env.sql
```SQL
select count(1) from gp_segment_configuration;
select dbid, content, role
  from gp_segment_configuration
  where status='d'
  order by dbid;
```
An expected answer file is under `expected/`. If we run
`./pg_regress  ... --schedule=./xxx_schedule --hook-test=check_env`, pg_regress will run check_env after every test case.
We could add another option `--hook-points` to tell pg_regress/pg_isolation2_regress to run the check test only after a set of test cases.
hook.txt
```
# comment
setup
createdb
```
`./pg_regress  ... --schedule=./xxx_schedule --hook-test=check_env --hook-points=hook.txt`
Now, pg_regress will run the check(check_env) only after the test cases, `setup` and `createdb`.

**Comment by asimrp** [comment]
A test should cleanup after itself and leave the environment in a sane state, so that subsequent tests are not affected.  If a test has passed but left the cluster in a bad state, then there is bug somewhere.  The change proposed in this PR will hide such bugs.

**Comment by gfphoenix78** [comment]
> A test should cleanup after itself and leave the environment in a sane state...

Yes, everyone should be a good man, it's just a beautiful wish. We must be able to cope with some evil guys.
Besides, it's difficult and time-consuming to find out the bad guys. Once a bad test case added, how/when do we find it out?

**Comment by asimrp** [comment]
@gkokolatos and I talked about something like this while analyzing test failures in PG12 merge branch.  We have lots of tests in isolation2 suite that verify the cluster state as first step.  But even when the verification fails, there is no way to stop running commands from the test case.  It would be nice to have a way within a testcase to signal the test harness to stop when the cluster is not in the expected state.

File: src/test/regress/pg_regress.c [hunk 1]
```diff
@@ -1990,6 +2069,70 @@ log_child_failure(int exitstatus)
...
+	wait_for_tests(&pids, &statuses, NULL, NULL, 1);
+
+	bool diff = results_differ(testname, resultfiles->str, expectfiles->str);
+	free_stringlist(&resultfiles);
+	free_stringlist(&expectfiles);
+	free_stringlist(&tags);
+	if (!diff)
+		return;
+
+	/* report error */
```
Comment (asimrp) [inline src/test/regress/pg_regress.c:2131]: No more tests should be run after this point.

**Comment by gfphoenix78** [comment]
> 1. Did we evaluate the option to check cluster status at the end of each test?  If it doesn't cause significant increase in run time, the `--hook*` flags can be eliminated.

Yes, we could. Checking the cluster status will run after each test or a set of specific test cases. If we just want to run the check script after some suspicious test cases, we could only run it several times. It depends on how we want to narrow down the guilty test cases.

> 2. The test runner (pg_regress or isolation2) should exit without running any more tests if the state of the cluster is not clean.  There is no point in running tests against an unclean cluster, the results are not trustworthy.

Good point. It makes sense.

> 3. The SQL commands to cluster health check should contain:
> 
> ```sql
> select gp_inject_fault('all', 'reset', dbid) from gp_segment_configuration;
> select * from gp_request_fts_probe_scan();
> select * from gp_segment_configuration where status = 'd' or mode = 'n';
> ```
> 
> Let's see how much overhead these SQL statements incur and if it is too much (in terms of test run times), the probe scan step can be skipped.

It may be time-consuming to run the probe. I think it's much quicker to count the number of remaining fault-injector items, if it is a problem.

**Comment by asimrp** [comment]
The idea of introducing new command line options `--hook*` is not particularly appealing.  We should try to achieve the same functionality with much less disruption.  How about treating a test name in a special way?  E.g. "stop_if_cluster_not_clean".  We can add `test: stop_if_cluster_not_clean` lines to schedule files as and when needed.  The test runner recognizes this special test name and if that test fails, it exits without running any further tests.

**Comment by gfphoenix78** [comment]
@asimrp I don't think it is a good idea to use a special test name. The hook script is accidentally used to implement checking the cluster status, but it can do much more. If you concern adding an additional option to pg_regress/isolation2_pg_regress, I could keep only one option, a configuration file, which specifies the test script file, running script file after each test, or a set of tests.

**Comment by gfphoenix78** [comment]
@asimrp  Now, pg_regress has only one parameter for the hook in this PR, `--hook-file`. It specifies the hook configuration file. The file format is:
```
# comment
# the check_cluster is the test name that we want to run
script: check_cluster

# type: specifies which type of hook we want to run. The valid type names are:
# 'all', 'subset'.
# 'all' means to run the hook after each test/test group.
# 'subset' means we just want to run the hook after a set of tests which is followed line by line.
# 
# type: all
# type: subset

# If the type is 'subset', the following lines are the set of test names, one test name per line.
# boolean
# tablespace
```

Is it fine to add the hook in this PR and put/update the script to check the cluster status in another PR? Checking the remaining fault injector items is necessary, but I don't think we should run FTS probe.
Of course, we may change the title of this PR.

**Comment by zhangh43** [comment]
Thanks @gfphoenix78  for pointing out this issue. In fact, when handling gprecoverseg test case, I'm also struggled with the `wait for cluster to be sync` issue. Sometimes we use utility mode to connect a segmentDB to check the failover succeed or not, sometimes we use a loop to check the gp_segment_configuration state. I think it's very helpful to wrap it in, e.g., `isolation2/helpers/server_helpers.sql` for isolation2 test. As a result, we could use a unified way to check the cluster status.

This could achieve the similar effect as the 'hook*' proposed in this PR. Committer who add the test case could just call a UDF to check the cluster in a unified way. Also note that this is not only useful between test cases, but also inside the test case itself.

The test case, which call this UDF, should also not be grouped with others, since it may change the cluster status. We should note it in the README.

For aborting the test when cluster is not synced, I think it's nice to have feature. Since we already have enough clue in the diff file of the first failed case, which indicates the cluster is not in sync.

File: src/test/regress/pg_regress.c (hunk 1 above)
Comment (gfphoenix78) [reply src/test/regress/pg_regress.c:2166]: Updated it.

**Comment by zhangh43** [comment]
The option `hook-test` and `hook-point` are not used by the ICW yet. I think it's better to specify the test cases which should use this hook, which should not. And enable this option in ICW.

**Comment by gfphoenix78** [comment]
> The option `hook-test` and `hook-point` are not used by the ICW yet. I think it's better to specify the test cases which should use this hook, which should not. And enable this option in ICW.

There is only one option `--hook-file` now.

**Comment by zhangh43** [comment]
> > The option `hook-test` and `hook-point` are not used by the ICW yet. I think it's better to specify the test cases which should use this hook, which should not. And enable this option in ICW.
> 
> There is only one option `--hook-file` now.

Got it, please update the commit message.  Currently the hook-file is empty and ICW will not utilize this feature, right?

File: src/test/regress/pg_regress.c [hunk 2]
```diff
@@ -1990,6 +2105,73 @@ log_child_failure(int exitstatus)
...
+	}
+	return NULL;
+}
+
+static void
+run_post_hook(const char *testname, test_function tfunc, const char **hook_points_array, int n_array, const char **tests, int n_tests)
+{
+	if (testname == NULL)
+		return;
+	const char *matched_test;
```
Comment (zhangh43) [inline src/test/regress/pg_regress.c:2137]: variable declaration should at the beginning of function

**Comment by gfphoenix78** [comment]
> > > The option `hook-test` and `hook-point` are not used by the ICW yet. I think it's better to specify the test cases which should use this hook, which should not. And enable this option in ICW.
> > 
> > 
> > There is only one option `--hook-file` now.
> 
> Got it, please update the commit message. Currently the hook-file is empty and ICW will not utilize this feature, right?

No, the hook file is not empty. The only one test is check_cluster, which is the hook to run after every test cases/group in regress and isolation2.

**Comment by asimrp** [comment]
A couple of suggestions came out of offline discussions, let me summarize them here:
1. Let's measure difference in ICW runtime if the proposed cluster check is performed after every test.
2. If the difference in option 1 is unacceptable, leverage the ability to specify multiple schedule files on `pg_regress` command line.  The tests that need cluster check can be separated into a new schedule file and `pg_regress` can be taught to run cluster check after every test in this new schedule file.
3. One problem with this proposal is having to list tests that need cluster check in a hook file when they are also listed in a schedule file.  This seems confusing, we should avoid listing the same tests more than once.

**Comment by gfphoenix78** [comment]
> A couple of suggestions came out of offline discussions, let me summarize them here:
> 
> 1. Let's measure difference in ICW runtime if the proposed cluster check is performed after every test.

Yes, will do, but not on GCP. Let me think about it.

> 2. If the difference in option 1 is unacceptable, leverage the ability to specify multiple schedule files on `pg_regress` command line.  The tests that need cluster check can be separated into a new schedule file and `pg_regress` can be taught to run cluster check after every test in this new schedule file.
> 3. One problem with this proposal is having to list tests that need cluster check in a hook file when they are also listed in a schedule file.  This seems confusing, we should avoid listing the same tests more than once.

It's not a good idea to reorder the tests and split the tests from schedule files into a new file in order to run the cluster check.
Reordering the tests is not necessary and may be time-consuming because of dependencies. The set of tests that needs to run the cluster check may change several times depending on the test results.
I'm not feeling good to mix running the hook test and the normal tests. I prefer to decouple them.

**Comment by asimrp** [comment]
@gfphoenix78 did we get around to measuring the overhead of enabling cluster check after every `test:` line processed by the test runner?  This PR would avoid builds getting stuck like this: https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb_master/jobs/icw_gporca_ictcp_centos7/builds/248  It is stuck for over 24 hours because `fts_errors` test failed and it caused several subsequent tests to fail and one of them is stuck (I didn't investigate why).

**Comment by gfphoenix78** [comment]
> @gfphoenix78 did we get around to measuring the overhead of enabling cluster check after every `test:` line processed by the test runner? This PR would avoid builds getting stuck like this: https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb_master/jobs/icw_gporca_ictcp_centos7/builds/248 It is stuck for over 24 hours because `fts_errors` test failed and it caused several subsequent tests to fail and one of them is stuck (I didn't investigate why).

It's a long time to apply a VM in VMware. I'll have a test on the Pivotal VM today. Thank you for your focus.

**Comment by ashwinstar** [comment]
On Wed, Aug 5, 2020 at 8:25 PM Hao Wu <notifications@github.com> wrote:

>
>
>
>
> @gfphoenix78 <https://github.com/gfphoenix78> did we get around to
> measuring the overhead of enabling cluster check after every test: line
> processed by the test runner? This PR would avoid builds getting stuck like
> this:
> https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb_master/jobs/icw_gporca_ictcp_centos7/builds/248
> It is stuck for over 24 hours because fts_errors test failed and it
> caused several subsequent tests to fail and one of them is stuck (I didn't
> investigate why).
>
>
>
>
> It's a long time to apply a VM in VMware. I'll have a test on the Pivotal
> VM today. Thank you for your focus.
>

What’s stopping you from running tests locally on laptop or workstation and
getting the numbers. I periodically run Regress test of ICW on laptop fine
in 30mins or so.

**Comment by gfphoenix78** [comment]
The content of the test runner is:
```SQL
select dbid, content, status, role, preferred_role, mode from gp_segment_configuration where status = 'd' or preferred_role != role;
select gp_inject_fault('all', 'reset', dbid) from gp_segment_configuration where role='p';
```

Here are the results:
isolation2 on machine A (VM in shanghai)
| Without posthook  | With posthook |
| ------------- | ------------- |
| 9m58.278s  | 10m48.278s  |
| 10m16.482s  | 10m59.325s  |
|10m7.144s | 10m50.999s |
|10m11.444s | 10m42.237s |
|10m15.263s | 10m46.152s |

regress test on machine A (VM in shanghai)
| Without posthook  | With posthook |
| ------------- | ------------- |
| 20m6.670s | 21m50.869s |
| 21m24.760s | 22m0.185s |
| 21m17.381s | 659m4.829s |
| 21m42.749s | 1235m36.547s |
| 21m54.906s | 834m6.842s |

The last 3 values with posthook should be an exception.

isolation2 test on machine B (GCP, failed to get the result for regress with posthook)
| Without posthook  | With posthook |
| ------------- | ------------- |
| 12m51.753s  | 12m16.513s  |
| 12m21.505s  | 12m22.005s  |
|11m41.165s | 12m9.547s |
|11m48.940s | 12m22.161s |
|11m54.711s | 227m4.565s |

The last value with posthook should be an exception.
I also run the whole ICW tests in the pipeline, with or without the test runner, the difference time is little.

I'm not sure whether the exception data is caused by configuration options, machine-specific, or anything else.

**Comment by asimrp** [comment]
Thank you @gfphoenix78 for measuring the times.  I also feel that the difference is not significant and the check should be performed after every `test:` line in the schedule file.

**Comment by gfphoenix78** [comment]
Yes, we could run the test runner for all test cases. Currently, we need to check the role and status of all segments(including the master/standby). Another thing we must take care of is the fault injector. Since we don't have a way to check whether some fault injector entries remain in a segment, we could simply reset all of them on the primary if the segment is up.
`select gp_inject_fault('all', 'reset', dbid) from gp_segment_configuration where role='p' and status='u';`
Any ideas?

**Comment by ashwinstar** [comment]
Is it possible to easily do this check only for greenplum schedule and avoid it for upstream parallel schedule tests? As for sure we don’t need it for them.

Also, I am not sure why fault injector needs to be reset? Ideally, check should be read-only operation only. We should leave that to individual test to do. If fault is not reset, anyways should be caught by pipeline easily. Given even on failures all sqls are executed from file, faults should get reset for failing test as well.

**Comment by gfphoenix78** [comment]
> Is it possible to easily do this check only for greenplum schedule and avoid it for upstream parallel schedule tests? As for sure we don’t need it for them.

It needs some effort, not too much.

> 
> Also, I am not sure why fault injector needs to be reset? Ideally, check should be read-only operation only. We should leave that to individual test to do. If fault is not reset, anyways should be caught by pipeline easily. Given even on failures all sqls are executed from file, faults should get reset for failing test as well.

@ashwinstar I would like to only check the number of remaining fault injector entries in the shared memory, but fault injector doesn't provide such functions. Whether to add this checking function needs further discussion. Fault injector is also part of the cluster status, which is assumed to be clean before running the test case. If the current test case remains a fault injector entry in the shared memory, then the next test case may fail because of unpredictable cluster status. This is what the PR wants to do.

**Comment by ashwinstar** [comment]
On Wed, Aug 12, 2020 at 6:48 PM Hao Wu <notifications@github.com> wrote:

> Also, I am not sure why fault injector needs to be reset? Ideally, check
> should be read-only operation only. We should leave that to individual test
> to do. If fault is not reset, anyways should be caught by pipeline easily.
> Given even on failures all sqls are executed from file, faults should get
> reset for failing test as well.
>
> @ashwinstar <https://github.com/ashwinstar> I would like to only check
> the number of remaining fault injector entries in the shared memory, but
> fault injector doesn't provide such functions. Whether to add this checking
> function needs further discussion. Fault injector is also part of the
> cluster status, which is assumed to be clean before running the test case.
> If the current test case remains a fault injector entry in the shared
> memory, then the next test case may fail because of unpredictable cluster
> status. This is what the PR wants to do.
>

I think it’s different for fault injector compared to cluster status.
Cluster can get in not sane state due to test Fts triggering or failure or
recoverseg not working and such. Which won’t hit during development or
unpredictable. Whereas if test misses to reset the fault, it affecting some
other test will be detected during (local) test run before committing
itself.

I want us to write code only if needed with cases encountered and not on
speculations.  Do we have current case from pipeline where fault from one
test not reset affected other test?

**Comment by gfphoenix78** [comment]
> On Wed, Aug 12, 2020 at 6:48 PM Hao Wu ***@***.***> wrote: Also, I am not sure why fault injector needs to be reset? Ideally, check should be read-only operation only. We should leave that to individual test to do. If fault is not reset, anyways should be caught by pipeline easily. Given even on failures all sqls are executed from file, faults should get reset for failing test as well. @ashwinstar <https://github.com/ashwinstar> I would like to only check the number of remaining fault injector entries in the shared memory, but fault injector doesn't provide such functions. Whether to add this checking function needs further discussion. Fault injector is also part of the cluster status, which is assumed to be clean before running the test case. If the current test case remains a fault injector entry in the shared memory, then the next test case may fail because of unpredictable cluster status. This is what the PR wants to do.
> I think it’s different for fault injector compared to cluster status. Cluster can get in not sane state due to test Fts triggering or failure or recoverseg not working and such. Which won’t hit during development or unpredictable. Whereas if test misses to reset the fault, it affecting some other test will be detected during (local) test run before committing itself. I want us to write code only if needed with cases encountered and not on speculations. Do we have current case from pipeline where fault from one test not reset affected other test?

Yes, I saw that resetting fault injector failed in my local test. The fault injector name is `finish_prepared_start_of_function`. After this failure, the DTX recovery can't finish. I'm not sure why it failed, but it really happened. If the fault injector is not checked or unexpected, we don't know why the next test cases failed. Does it fail by the test itself or by unexpected status? This PR wants to ensure that if a test(except the first one) in regress/isolation2 fails, it must be caused by the test itself, not by the cluster status(including fault injector).

**Review by asimrp — COMMENTED** [review]
I feel that this PR is a bit over-engineered.  Based on the performance measurements, we know that the overhead incurred by checking cluster status is minimal and it can be done after each test run.  Such a functionality can be implemented with much less code change than what's being proposed here.  In particular, the newly introduced hook file and the code to parse it seems unnecessary.

As an example, I would like to propose this alternative implementation: https://github.com/asimrp/gpdb/commit/a91256e233a1814cae3396f8559b738979b7265f

**Comment by paul-guo-** [comment]
I roughly looked at the code. I agree with Asim's code which is simpler to resolve the pain point. We should not spend too much efforts/designs on testing itself.

Also my experience is that this probblem mainly exists in isolation2 tests (correct me if I'm wrong) so probably enable it in isolation2 test Makefile only. Also we could  remove some existing cluster status checking sql in isolation2 tests.

**Comment by gfphoenix78** [comment]
@asimrp It looks good to me. Go ahead, I'll close this PR in favor of your patch.

**Comment by gfphoenix78** [comment]
One thing I care about is the fault injector. If the previous tests inject some entries and failed to reset them, the remaining fault injector could be harmful to the next tests. It could be another story to check the number of entries for fault injector.

**Comment by asimrp** [comment]
PR #10825 is opened with the simplified change, as per Wu Hao's suggestion:

**Pull request closed without merging**